### ISSUE TITLE
We should remove all ActiveRel query methods

### ISSUE THREAD

**Issue by subvertallchris** [neo4jrb/neo4j#809]
There's never any good reason to use them, IMO, and I have no interest in maintaining them. Everything you can do with them can be done safer and with more flexibility from ActiveNode models in QueryProxy or Core::Query.

**Comment by pdipietro** [comment]
Hmmmm, little perplexity.

IMHO, I found it is a little confusing (and maybe not clearly documented) the differents between the reasons to use more than a single Query model (Query::Proxy, Core::Query, etc).

By a final user point of view, there should be just one single way to do queries, not needing to make tech transforms lake as-query or similar.

So please, before removing ActiveRel, should be better to approach the above points.

**Comment by cheerfulstoic** [comment]
I agree that the distinction between `QueryProxy` and `Query` should be more well defined.  Chris has a good point about it that I think should be copied into our new documentation:

http://neo4jrb.io/blog/2015/02/08/query_or_query-proxy.html

Though I don't think Chris is proposing removing `ActiveRel`.  He's just saying that we should remove the query methods on `ActiveRel` because the Neo4j paradigm is to start with nodes, not relationships

**Comment by subvertallchris** [comment]
Right, we are definitely not removing ActiveRel. It's just that it has its
own basic query methods ('all', 'first', "where", maybe a few others) that
are extremely basic, limited, have the potential to be terribly slow, and
essentially unmaintained. I want to remove those because they do not add
anything new, they just try to copy what you can already do better with
other classes.

**Comment by subvertallchris** [comment]
Any objections to this? Still think it's a good idea before the 5.0 release. I can take care of it this week.

**Comment by cheerfulstoic** [comment]
Yeah, I would say to remove them.  Definitely make a note in the `CHANGELOG.md` so that people can note it

**Comment by pdipietro** [comment]
I agree, remove them.

**Comment by mmmmmrob** [comment]
One of the reasons to love Neo4J (and property graph databases generally) is that nodes and edges have equal prominence. I'd like to see ActiveRel provide equivalent access to the graph as ActiveNode. find_or_create_by is an obvious missing for me when, for example, incrementing weights on edges.

**Comment by subvertallchris** [comment]
Prominent as they are, they can't be viewed as absolute equals with the
same abilities. Neo4j best practices insist against treating relationships
like nodes, which I take to mean avoiding overuse of properties and using
them for the start/end of a match. ActiveRel might be able to aid in the
pursuit of behavioe but nodes and QueryProxy can (and IMO should) do a
better job of working with Cypher.

Regarding find_or_create_by, maybe open an issue about that, we'll
brainstorm on how to do it?

**Comment by djvs** [comment]
We use these in the cases where we use an ActiveRel to optimize for querying relations, but want to store some (future-updatable) data on the relationship.  They are inefficient to do .find(..) on, so we never use them in loops, but we do use .find on them once in a while, at a cost of a few hundred ms.  We also retrieve them via pluck in query_as from their associated nodes.

**Comment by subvertallchris** [comment]
Could you elaborate on that first part?

**Comment by djvs** [comment]
Where we don't want a two-hop instead of a one-hop.  A user permissioning structure, a polymorphic tree structure, stuff like that.

**Comment by subvertallchris** [comment]
Been a while but coming back to this. We are going to go ahead and deprecate the entire `Neo4j::ActiveRel::Query` module. It's so small, about 90 lines, and the few things it offers can be rewritten locally within an app if there's no other alternative. I'm happy to help anyone who needs to do this.

**Comment by subvertallchris** [comment]
This will be done in 6.0. We're going to keep the basic `find`.

**Comment by subvertallchris** [comment]
We deprecated them in 6.0.0.alpha.1, so I'll leave this open until the 7.0 release as a reminder to remove them then.

**Comment by djvs** [comment]
I checked and we are only using .find by the way.
